### ISSUE TITLE
Original values are not updated after a write

### ISSUE THREAD

**Issue by NightJar** [silverstripe/silverstripe-framework#8443]
# TL;DR
A `DataObject` should have the same state after a `write` that it would having been freshly `get`

## Affected Version

4.3.x-dev 3b7802bb51ee32432ed8b45a722275f5243ffcbf
Although probably affecting versions back near enough to the [beginning of time](https://github.com/silverstripe/silverstripe-framework/blob/2.4/core/model/DataObject.php).

## Description

Changes are tracked in two ways; one by updating a `changed` private array, and another via a copy of `record` kept in the field `original`, set at `__construct` - both ways can be accessed via the `getChangedFields` method.
After calling `write` one would expect to have the local instance of the object updated to match that which would be produced by a brand new query for it via the active record controls (`get`). However this is not the case, and one cannot, for example, look up the value a particular field had before it was changed in order to reset it in `onBeforeWrite`.

## Steps to Reproduce

1. Code a `DataObject` subclass. It could be empty (no extra DB fields) - it's not important the content.
2. `dev/build`
3. `create` one of these objects
4. `update` the `Title`
5. `write`
6. `update` the `Title`
7. `write`
8. `update` the `Title`
9. `write`
10. `update` the `Title`
11. call `getChangedFields('Title')`
12. **behold the `before` value of `null`.**
13. use that existing object's ID to `get` a new instance of it from the database.
14. repeat step 10, but on the new instance.
15. repeat step 11, but on the new instance.
16. behold the `before` value of that which was saved at step 9.

**Comment by NightJar** [comment]
Hmm, seems like that was originally a duplicate yes, however got a bit off topic in dealing with filenames. Perhaps this is why @sminnee closed it, however he hasn't reported what it was he tested in order to do so (or reference a commit where this is fixed but yet to be released).

I can make a failing test, sure. But I'm afraid it shaln't be tonight :zzz:

**Comment by NightJar** [comment]
Ideal thanks @sminnee - I'll check it out and implement the test as @robbieaverill has asked.

**Comment by sminnee** [comment]
3821 is pretty similar to your test https://gist.github.com/svandragt/42863fdc18e7062955c0

**Comment by robbieaverill** [comment]
Ok I think we can close this as a duplicate now

**Comment by NightJar** [comment]
Sure. Two sides to the same coin really.
